File: 2000290120051/Queue.cpp
#include <iostream>

using std::cin;
using std::cout;
using std::endl;

class Queue {
    int *a;
    int front = 0, back = -1;
    int capacity;
    int size;
    
public:
    Queue(int k) {
        a = new int[k];
        front = 0;
        back = -1;
        size = 0;
        capacity = k;
    }

    ~Queue() { delete a; }

    bool isFull() { return size == capacity; }
    bool isEmpty() { return size == 0; }
    int getSize() { return size; }

    int getFront() {
        if(isEmpty()) {
            cout << "UNDERFLOW : Can't fetch front of Queue as Queue is Empty. RETURNING -1...." << endl;
            return -1;
        }

        return a[front];
    }
    int getBack() {
        if(isEmpty()) {
            cout << "UNDERFLOW : Can't fetch back of Queue as Queue is Empty. RETURNING -1...." << endl;
            return -1;
        }

        return a[back];
    }

    void enqueue(int val) {
        if(isFull()) {
            cout << "OVERFLOW : Value " << val << " can't be enqueued in queue as queue is full." << endl;
            return;
        }

        back = (back + 1) % capacity;
        ++size;
        a[back] = val;
    }

    void dequeue() {
        if(isEmpty()) {
            cout << "UNDER : Queue is empty." << endl;
            return;
        }

        front = (front + 1) % capacity;
        --size;
    }

    void printQueue() {
        if(isEmpty()) {
            cout << "UNDERFLOW : Can't print, Queue is empty." << endl;
            return;
        }

        int t = size, i = front;
        cout << "Front -> ";
        while(t) {
            cout << a[i] << " ";
            i = (i + 1) % capacity;
            --t;
        }
        cout << "<- Back" << endl;
    }
};

int main() {
    Queue q(10);
    
    cout << "Is queue empty : " << (q.isEmpty() ? "True" : "False") << endl;
    cout << "Is queue full : " << (q.isFull() ? "True" : "False") << endl;

    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);
    q.enqueue(40);
    q.enqueue(50);
    q.enqueue(60);
    q.enqueue(70);
    q.enqueue(80);
    q.enqueue(90);
    q.enqueue(100);
    q.enqueue(110);

    q.printQueue();
    cout << "Is queue empty : " << (q.isEmpty() ? "True" : "False") << endl;
    cout << "Is queue full : " << (q.isFull() ? "True" : "False") << endl;

    q.dequeue();
    q.dequeue();
    q.dequeue();
    q.dequeue();

    q.printQueue();
    cout << "Is queue empty : " << (q.isEmpty() ? "True" : "False") << endl;
    cout << "Is queue full : " << (q.isFull() ? "True" : "False") << endl;

    cout << endl << "Queue Size : " << q.getSize() << endl;
    cout << "Queue Front : " << q.getFront() << endl;
    cout << "Queue Back : " << q.getBack() << endl;

    return 0;
}

File: 2000290120054/Week_2_Stack_Queue/Queue.java
public class Queue {
    private int[] queue;
    private int front;
    private int rear;
    private int size;
    private int capacity;

    public Queue(int capacity) {
        this.capacity = capacity;
        queue = new int[capacity];
        front = 0;
        rear = -1;
        size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public int size() {
        return size;
    }

    public void enqueue(int item) {
        if (isFull()) {
            System.out.println("Queue is full. Cannot enqueue " + item);
            return;
        }
        rear = (rear + 1) % capacity;
        queue[rear] = item;
        size++;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue is empty. Cannot dequeue.");
            return -1;
        }
        int item = queue[front];
        front = (front + 1) % capacity;
        size--;
        return item;
    }

    public int front() {
        if (isEmpty()) {
            System.out.println("Queue is empty.");
            return -1;
        }
        return queue[front];
    }

    public int rear() {
        if (isEmpty()) {
            System.out.println("Queue is empty.");
            return -1;
        }
        return queue[rear];
    }

    public static void main(String[] args) {
        Queue queue = new Queue(5);

        queue.enqueue(10);
        queue.enqueue(20);
        queue.enqueue(30);

        System.out.println("Front: " + queue.front());
        System.out.println("Rear: " + queue.rear());

        System.out.println("Dequeued item: " + queue.dequeue());

        queue.enqueue(40);
        queue.enqueue(50);
        queue.enqueue(60);

        while (!queue.isEmpty()) {
            System.out.println("Dequeued item: " + queue.dequeue());
        }
    }
}
